### PR TITLE
Verbose logging

### PR DESCRIPTION
**Pull request by george-ayris**
By default octopose now only outputs logs for subprocceses when they return a non-zero exit code (to help with debugging). To get all logs you can use the `-v` or `--verbose` flag.

Solves #14 and #13

### CODE REVIEW

File: util.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# MIT License
+#
+# Copyright (c) 2017 Huddle
```
Comment (domgreen) [inline util.py]: Can probably change the date :p

File: util.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import sys
+
+
+def run_subprocess(args, error_msg, verbose):
+    completed_process = subprocess.run(args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=False)
+
+    if completed_process.returncode != 0:
+        print(completed_process.stdout.decode('utf-8'))
+        print(error_msg, file=sys.stderr)
+        exit()
```
Comment (domgreen) [inline util.py]: Do we always want to exit if a process fails? May want to continue and just report on the errors ... just a thought.

File: octopose.py [hunk 3]
```diff
@@ -27,21 +27,21 @@
...
 import os
-import subprocess
 import sys
 import octo
 import nu
+import util
 
 
-def invoke_deploy(step_path):
+def invoke_deploy(step_path, verbose):
```
Comment (domgreen) [inline octopose.py]: Is there a way to do this without having to pass `verbose` to every method? and jsut have `run_subprocess` know from the start if it should be verbose or not?

**Review by domgreen — COMMENTED** [review]
Looks good, just some comments around if we need to pass `verbose` in each method and if we always want to exit on error.

File: util.py (hunk 2 above)
Comment (george-ayris) [reply util.py]: After a little bit of discussion with Ian, we think the behvaiour that makes the most sense is that deploying multiple projects shouldn't cause each other to fail. However, within a project the order of packages is usually important, so if a `*Deploy.ps1` fails within a project then it shouldn't deploy any more packages/run any more scripts? 

If you agree, I might raise another issue for this.

File: octopose.py (hunk 3 above)
Comment (george-ayris) [reply octopose.py]: Unsure if I've gone a little too much towards classes with the refactoring, but I think there's now a slightly nicer separation of code. Interested to hear your thoughts.

**Comment by penderi** [comment]
I Like This

File: util.py (hunk 2 above)
Comment (domgreen) [reply util.py]: Yeah if one step fails fail that whole deployment but go on to process the rest of the deployments. 
Think it gets super annoying when everything fails cos some rookie in Babble breaks their section 🤪 agree do that on a separate PR

File: nu.py [hunk 4]
```diff
@@ -22,17 +22,16 @@
...
-def get_deployable(name, version, staging_location):
-    """ Get deployables from pacakage sources for local deployment. """
-    for source in config.PACKAGE_SOURCES:
-        FNULL = open(os.devnull, 'w')    #use this if you want to suppress output to stdout from the subprocess
-        args = "third_party\\NuGet.exe install {0} -Source {1} -OutputDirectory {2}".format(name, source, staging_location)
-        if version is not None:
-            args = args + " -Version {0}".format(version)
-        subprocess.call(args, stdout=FNULL, stderr=FNULL, shell=False)
+class Nu:
+    def __init__(self, subprocess_runner):
```
Comment (domgreen) [inline nu.py:29]: Can you do ctor style comments on this?

File: octopose.py [hunk 5]
```diff
@@ -169,7 +61,8 @@ def deploy_local(data):
         manifest_string = infile_contents
 
     manifest = json.loads(manifest_string)
+
     if environment != "local":
-        deploy_to_environment(env_id, wait, force, manifest)
+        remote_deploy.deploy_to_environment(env_id, wait, force, manifest)
     else:
-        deploy_local(manifest)
+        local_deploy.LocalDeploy(verbose).deploy(manifest)
```
Comment (domgreen) [inline octopose.py]: `local_deploy` `localDeploy` `deploy` ... does this per chance do a deploy?

**Review by domgreen — COMMENTED** [review]
Not a huge fan of OO python for such a small case but looks legit.

My initial thought would be something as hacky as setting something global (i know its still dirty) for the log level.
Or ... `local_deply.verbose(True)` `local_deploy.run(manifest)`

**Comment by george-ayris** [comment]
Have updated with some (hopefully) not controversial docstring comments. Going to leave in the OO approach and see how it feels (but will consider changing it back later).

**Pull request merged**